src/plugins/intel_gpu/src/runtime/ze/ze_engine.cpp

```diff
@@ -0,0 +1,236 @@
+
+        return res;
+    } catch (const std::exception& e) {
+        OPENVINO_THROW("[GPU] Failed to allocate memory: ", e.what());
+    }
+}
+
+memory::ptr ze_engine::reinterpret_buffer(const memory& memory, const layout& new_layout) {
+    OPENVINO_ASSERT(memory.get_engine() == this, "[GPU] trying to reinterpret buffer allocated by a different engine");
+    OPENVINO_ASSERT(new_layout.format.is_image() == memory.get_layout().format.is_image(),
+                    "[GPU] trying to reinterpret between image and non-image layouts. Current: ",
+                    memory.get_layout().format.to_string(), " Target: ", new_layout.format.to_string());
+
+    if (memory_capabilities::is_usm_type(memory.get_allocation_type())) {
+            return std::make_shared<ze::gpu_usm>(this,
+                                     new_layout,
+                                     reinterpret_cast<const ze::gpu_usm&>(memory).get_buffer(),
+                                     memory.get_allocation_type(),
+                                     memory.get_mem_tracker());
+    }
+
+    OPENVINO_THROW("[GPU] Trying to reinterpret non usm buffer");
+}
+
+memory::ptr ze_engine::reinterpret_handle(const layout& new_layout, shared_mem_params params) {
+    if (params.mem_type == shared_mem_type::shared_mem_usm) {
+        ze::UsmMemory usm_buffer(get_context(), get_device(), params.mem);
+        size_t actual_mem_size = 0;
+        zeMemGetAddressRange(get_context(), params.mem, nullptr, &actual_mem_size);
+        auto requested_mem_size = new_layout.bytes_count();
+        OPENVINO_ASSERT(actual_mem_size >= requested_mem_size,
+                            "[GPU] shared USM buffer has smaller size (", actual_mem_size,
+                            ") than specified layout (", requested_mem_size, ")");
+        return std::make_shared<ze::gpu_usm>(this, new_layout, usm_buffer, nullptr);
+    } else {
+        OPENVINO_THROW("[GPU] Unsupported shared memory type: ", params.mem_type);
+    }
+}
+
+memory_ptr ze_engine::create_subbuffer(const memory& memory, const layout& new_layout, size_t byte_offset) {
+    OPENVINO_ASSERT(memory.get_engine() == this, "[GPU] Trying to create a subbuffer from a buffer allocated by a different engine");
+    if (new_layout.format.is_image_2d()) {
+        OPENVINO_NOT_IMPLEMENTED;
+    }
+    if (memory_capabilities::is_usm_type(memory.get_allocation_type())) {
+        auto& new_buf = reinterpret_cast<const ze::gpu_usm&>(memory);
+        auto ptr = new_buf.get_buffer().get();
+        auto sub_buffer = ze::UsmMemory(get_context(), get_device(), ptr, byte_offset);
+        return std::make_shared<ze::gpu_usm>(this,
+                                 new_layout,
+                                 sub_buffer,
+                                 memory.get_allocation_type(),
+                                 memory.get_mem_tracker());
+    } else {
+        OPENVINO_THROW("[GPU] Trying to create subbuffer for non usm memory");
+    }
```

Comment:
I guess it can be simplified because does not support cl_mem type
```
OPENVINO_ASSERT(memory_capabilities::is_usm_type(memory.get_allocation_type());
auto & new_buf = re...
```